esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+        StringBuffer hostInPath, localPath, absPath;
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath);
+
+        //Validate file path.
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
+        if (containsRelPaths(path)) //Detect a path like: a/../../../f
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", path);
+
+        if (isAbsolutePath(path))
+        {
+            //Check if the path is valid.
+            if (!startsWith(path, dropZonePlanePath))
+                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid source path %s for spraying a file.", path);
         }
         else
         {
-            sourceIPReq.set(srcIP).trim();
-            if (sourceIPReq.isEmpty())
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            sourcePlaneReq.append(plane->queryProp("@name"));
+            //Insert the @prefix in the front of the file path
+            absPath.set(dropZonePlanePath);
+            addNonEmptyPathSepChar(absPath);
+            absPath.append(path);
+            path = absPath.str(); //Now, the path is an absolute path.
+        }
+
+        //If the hostInPath string is not empty (ex. '//10.0.0.1'), it (skipping the '//') should be used to check the scope permission.
+        ensureDZFileScopePermission(context, sourcePlaneReq, hostInPath.isEmpty() ? sourceHostReq : hostInPath + 2, path, SecAccess_Read);
+
+        if (!sourcePathReq.isEmpty())
+            sourcePathReq.append(",");
+        if (!hostInPath.isEmpty())
+            sourcePathReq.append(hostInPath);
+        sourcePathReq.append(path);
+    }
+
+    if (sourceHostReq.isEmpty())
+        getDropZoneHost(sourcePlaneReq, dropZone, sourceHostReq);
+}
+
+void CFileSprayEx::checkSpraySourcePathAndGetPlane(IEspContext& context, const char* sourceHost, StringArray& fileReqs,
+    StringBuffer& sourcePlaneReq)
+{
+    SocketEndpoint sourceHostEp(sourceHost);
+    bool firstFileReq = true;
+    Owned<IPropertyTree> dropZone;
+    ForEachItemIn(i, fileReqs)
+    {
+        const char* fileReq = fileReqs.item(i);
+        if (isEmptyString(fileReq))
+            continue;
+
+        //Parse the fileReq to find/validate possible host/ip and local path.
+        StringBuffer hostInPath, localPath;
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHost, &sourceHostEp, localPath, hostInPath);
+
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
+        if (containsRelPaths(path)) //Detect a path like: a/../../../f
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", fileReq);
+
+        if (firstFileReq)
+        {
+            firstFileReq = false;
```

Comment:
still seems to a logical problem here..
What happens if 2nd, 3rd UNC path have different hostname in them?

Looks like 1st will be validated, and all others ignored at the moment?
But judging by setSpraySourceDFUFileSpec and srcDFUfileSpec->setMultiFilename .. we don't support it anyway.. i.e. only 1 host has ever been permitted here.

If so, this should not ignore other UNC with different hosts, but throw an error and reject it

Reply:
I do assume all of the UNC paths have the same hostname now. If there are different hostnames, could they are in different planes? If so, which one should be set into the DFU WU or we throw an error?

Reply:
as I say, afaics, srcDFUfileSpec->setMultiFilename etc. wouldn't support it anyway, so should throw an error.

Reply:
I just realized that the checkHostInSpraySourcePath() does throw an error if hostnames are from different hosts. If a dropzone is found for the first file, the dropzone will be used for all files.

Reply:
The code should be revised to check the path also if a dropzone is found: if (!startsWith(path, dropZonePlanePath)) ...